File: src/Silver_2/BOJ7562.java
package Silver_2;

import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class BOJ7562 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());
        StringTokenizer st;
        int I = 0;
        StringBuilder sb = new StringBuilder();
        while (T-- > 0) {
            I = Integer.parseInt(br.readLine());

            int[] curPos = new int[2];
            st = new StringTokenizer(br.readLine());
            curPos[0] = Integer.parseInt(st.nextToken());
            curPos[1] = Integer.parseInt(st.nextToken());

            st = new StringTokenizer(br.readLine());
            int[] targetPos = new int[2];
            targetPos[0] = Integer.parseInt(st.nextToken());
            targetPos[1] = Integer.parseInt(st.nextToken());

            int[][] dist = new int[I+1][I+1];
            initDist(dist);

            if (curPos[0] == targetPos[0] && curPos[1] == targetPos[1]) {
                sb.append(0).append('\n');
            } else {
                bfs(dist, curPos, targetPos, I);
                sb.append(dist[targetPos[1]][targetPos[0]]).append('\n');
            }
        }

        System.out.println(sb);
    }

    private static void initDist(int[][] dist) {
        for(int i=0; i<dist.length; i++) {
            for (int j=0; j<dist.length; j++) {
                dist[i][j] = -1;
            }
        }
    }

    private static void bfs(int[][] dist, int[] curPos, int[] targetPos, int I) {
        Queue<int[]> queue = new LinkedList<>();
        queue.add(curPos);
        dist[curPos[1]][curPos[0]] = 0;

        int[] cur;
        int[] xPos = {-1, -1, 1, 1, -2, -2, 2, 2};
        int[] yPos = {2, -2, 2, -2, 1, -1, 1, -1};
        int dx = 0; int dy = 0;
        while (!queue.isEmpty()) {
            cur = queue.poll();

            for(int i=0; i<8; i++) {
                dx = cur[0] + xPos[i];
                dy = cur[1] + yPos[i];

                if (dx < 0 || dx >= I || dy < 0 || dy >= I) continue;
                if (dist[dy][dx] != -1) continue;
                dist[dy][dx] = dist[cur[1]][cur[0]] + 1;

                if (dx == targetPos[0] && dy == targetPos[1]) return;
                queue.add(new int[] { dx, dy });
            }
        }
    }
}

File: src/Silver_1/BOJ7562.kt
package Silver_1

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val case = br.readLine().toInt()
    val directX = intArrayOf(1, 2, 2, 1, -1, -2, -2, -1)
    val directY = intArrayOf(2, 1, -1, -2, -2, -1, 1, 2)

    repeat(case) {
        val q: Queue<Pair<Int, Int>> = LinkedList()
        val I = br.readLine().toInt()
        val dist = Array(I) { IntArray(I) }
        var st = StringTokenizer(br.readLine())
        val knightPos = Pair(st.nextToken().toInt(), st.nextToken().toInt())
        st = StringTokenizer(br.readLine())
        val wantPos = Pair(st.nextToken().toInt(), st.nextToken().toInt())

        if (knightPos == wantPos) {
            println("0")
            return@repeat
        }

        q.add(knightPos)
        while (q.isNotEmpty()) {
            val (curX, curY) = q.poll()

            for (i in 0 until 8) {
                val dx = curX + directX[i]
                val dy = curY + directY[i]

                if (dx < 0 || dx >= I || dy < 0 || dy >= I) continue
                if (dist[dy][dx] > 0) continue

                dist[dy][dx] = dist[curY][curX] + 1
                if (dx == wantPos.first && dy == wantPos.second) {
                    println(dist[dy][dx])
                    return@repeat
                }
                q.add(Pair(dx, dy))
            }
        }
    }
}